modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/CheckpointWorkflow.java

```diff
@@ -343,14 +365,35 @@ private DataRegionsDirtyPages beginCheckpoint(
             Collection<? extends DataRegion<PersistentPageMemory>> dataRegions,
             CompletableFuture<?> allowToReplace
     ) {
+        Map<DataRegion<?>, Set<FullPageId>> dirtyPartitionsMap = this.dirtyPartitionsMap;
+
+        this.dirtyPartitionsMap = new ConcurrentHashMap<>();
+
         Collection<DataRegionDirtyPages<Collection<FullPageId>>> dataRegionsDirtyPages = new ArrayList<>(dataRegions.size());
 
+        // First, we iterate all regions that have dirty pages.
         for (DataRegion<PersistentPageMemory> dataRegion : dataRegions) {
             Collection<FullPageId> dirtyPages = dataRegion.pageMemory().beginCheckpoint(allowToReplace);
 
+            Set<FullPageId> dirtyMetaPageIds = dirtyPartitionsMap.remove(dataRegion);
+
+            if (dirtyMetaPageIds != null) {
+                // Merge these two collections. There should be no intersections.
+                dirtyPages = CollectionUtils.concat(dirtyMetaPageIds, dirtyPages);
```

Comment:
I remember that you didn’t like `CollectionUtils#concat`, I see that you used it)

Reply:
Please take the context into account. In this case we must do it to avoid excessive allocations. `dirtyPages` is itself a concatenation.
Maybe streams would fit better, but the idea is the same

Reply:
Just a funny observation)